core/src/main/java/com/datastax/oss/driver/internal/core/cql/DefaultPagingState.java

```diff
@@ -0,0 +1,160 @@
+/*
+ * Copyright DataStax, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.datastax.oss.driver.internal.core.cql;
+
+import com.datastax.oss.driver.api.core.cql.BatchStatement;
+import com.datastax.oss.driver.api.core.cql.BoundStatement;
+import com.datastax.oss.driver.api.core.cql.PagingState;
+import com.datastax.oss.driver.api.core.cql.SimpleStatement;
+import com.datastax.oss.driver.api.core.cql.Statement;
+import com.datastax.oss.driver.api.core.detach.AttachmentPoint;
+import com.datastax.oss.driver.api.core.session.Session;
+import com.datastax.oss.driver.internal.core.data.ValuesHelper;
+import com.datastax.oss.protocol.internal.util.Bytes;
+import edu.umd.cs.findbugs.annotations.NonNull;
+import edu.umd.cs.findbugs.annotations.Nullable;
+import java.nio.ByteBuffer;
+import java.nio.charset.Charset;
+import java.security.MessageDigest;
+import java.security.NoSuchAlgorithmException;
+import java.util.Arrays;
+
+public class DefaultPagingState implements PagingState {
+
+  private final ByteBuffer rawPagingState;
+  private final byte[] hash;
+  private final int protocolVersion;
+
+  public DefaultPagingState(
+      ByteBuffer rawPagingState, Statement<?> statement, AttachmentPoint attachmentPoint) {
+    this(
+        rawPagingState,
+        hash(statement, rawPagingState, attachmentPoint),
+        attachmentPoint.getProtocolVersion().getCode());
+  }
+
+  private DefaultPagingState(ByteBuffer rawPagingState, byte[] hash, int protocolVersion) {
+    this.rawPagingState = rawPagingState;
+    this.hash = hash;
+    this.protocolVersion = protocolVersion;
+  }
+
+  // Same serialized form as in driver 3:
+  //     size of raw state|size of hash|raw state|hash|protocol version
+  //
+  // The protocol version might be absent, in which case it defaults to V2 (this is for backward
+  // compatibility with 2.0.10 where it is always absent).
+  public static DefaultPagingState fromBytes(byte[] bytes) {
+    ByteBuffer buffer = ByteBuffer.wrap(bytes);
+    short rawPagingStateLength = buffer.getShort();
+    short hashLength = buffer.getShort();
+    if (rawPagingStateLength + hashLength != buffer.remaining()
+        && rawPagingStateLength + hashLength + 2 != buffer.remaining()) {
```

Comment:
Is this `+ 2` a case of absent protocol version?

Reply:
Yes, we test two different length because the protocol version may or may not be there.

Reply:
I've rearranged the code to clarify this.